pkg/auth/oauth2authenticator.go

```diff
@@ -0,0 +1,252 @@
+	_ = server.Shutdown(context.Background())
+}
+
+func getRedirectUri(port int) string {
+	callback := fmt.Sprintf("http://%s:%d%s", CALLBACK_HOSTNAME, port, CALLBACK_PATH)
+	return callback
+}
+
+func getOAuthConfiguration(config configuration.Configuration) *oauth2.Config {
+
+	appUrl := config.GetString(configuration.WEB_APP_URL)
+	// tokenUrl := strings.Replace(appUrl, "app.", "id.", 1) + "/oauth2/default/v1/token"
+	tokenUrl := config.GetString(configuration.OAuthTokenUrl) // "https://snyk-fedramp-alpha.okta.com/oauth2/default/v1/token"
+	authUrl := appUrl + "/oauth/authorize"
+
+	conf := &oauth2.Config{
+		ClientID: OAUTH_CLIENT_ID,
+		Endpoint: oauth2.Endpoint{
+			TokenURL: tokenUrl,
+			AuthURL:  authUrl,
+		},
+	}
+	return conf
+}
+
+func getCodeChallenge(verifier []byte) string {
+	shasum := sha256.Sum256(verifier)
+	return base64.RawURLEncoding.EncodeToString(shasum[:])
+}
+
+func createVerifier(count int) []byte {
+	/*
+	  unreserved = ALPHA / DIGIT / "-" / "." / "_" / "~"
+	   ALPHA = %x41-5A / %x61-7A
+	   DIGIT = %x30-39
+	*/
+	lut := []byte("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~")
+	verifier := make([]byte, count)
+
+	// TODO is this good enough?
+	for i := range verifier {
+		index := rand.Int() % len(lut)
+		verifier[i] = lut[index]
+	}
+
+	return verifier
+}
+
+func getToken(config configuration.Configuration) (*oauth2.Token, error) {
+	oauthTokenString := config.GetString(CONFIG_KEY_OAUTH_TOKEN)
+	if len(oauthTokenString) > 0 {
+		token := &oauth2.Token{}
+		err := json.Unmarshal([]byte(oauthTokenString), token)
+		if err != nil {
+			return nil, err
+		}
+		return token, nil
+	}
+	return nil, nil
+}
+
+func NewOAuth2Authenticator(config configuration.Configuration, httpClient *http.Client) Authenticator {
+	token, _ := getToken(config)
+	oauthConfig := getOAuthConfiguration(config)
+
+	return &oAuth2Authenticator{
+		httpClient:         httpClient,
+		config:             config,
+		oauthConfig:        oauthConfig,
+		token:              token,
+		openBrowserFunc:    openBrowser,
+		shutdownServerFunc: shutdownServer,
+	}
+}
+
+func (o *oAuth2Authenticator) persistToken(token *oauth2.Token) {
+	tokenstring, _ := json.Marshal(token)
+	o.config.Set(CONFIG_KEY_OAUTH_TOKEN, string(tokenstring))
+	o.token = token
+}
+
+func (o *oAuth2Authenticator) Authenticate() error {
+	var responseCode string
+	var responseState string
+	var responseError string
+	verifier := createVerifier(128)
+	state := string(createVerifier(15))
+	codeChallenge := getCodeChallenge(verifier)
+	ctx := context.Background()
+
+	if o.headless {
+		// TODO:
+	} else {
+		srv := &http.Server{}
+
+		http.HandleFunc(CALLBACK_PATH, func(w http.ResponseWriter, r *http.Request) {
+			responseError = html.EscapeString(r.URL.Query().Get("error"))
+			if len(responseError) > 0 {
+				details := html.EscapeString(r.URL.Query().Get("error_description"))
+				fmt.Fprintf(w, "Error during authentication! (%s)\n%s", responseError, details)
```

Comment:
should this maybe use a logging framework?

Reply:
This is writing to the local web page that is showing either authentication successful or an error message. We actually need to check with Product if we need a more/designed webpage here.
But yes, in addition, we could log this into the debug logs.

Reply:
I just checked, we return this error if it occurs and it can be logged by the caller